csc/2018/2.LockFreeSet/GontaKA/LockFreeSetImpl.java

```diff
@@ -0,0 +1,102 @@
+import java.util.concurrent.atomic.AtomicMarkableReference;
+
+public class LockFreeSetImpl<T extends Comparable<T>> implements LockFreeSet<T> {
+    private class Node {
+        private final T value;
+        private final AtomicMarkableReference<Node> currentFlagAndNext;
+
+        Node(T value, Node next) {
+            this.value = value;
+            currentFlagAndNext = new AtomicMarkableReference<>(next, false);
+        }
+    }
+
+    private class Pair {
+        Node currentNode;
+        Node nextNode;
+
+        Pair(Node currentNode, Node nextNode) {
+            this.currentNode = currentNode;
+            this.nextNode = nextNode;
+        }
+    }
+
+    private final Node head = new Node(null, null);
+
+    public boolean add(T value) {
+        while (true) {
+            Pair pair = find(value);
+            Node currentNode = pair.currentNode;
+            Node nextNode = pair.nextNode;
+            if (nextNode != null && nextNode.value.compareTo(value) == 0) {
+                // Element already exists
+                return false;
+            } else {
+                Node node = new Node(value, nextNode);
+                if (currentNode.currentFlagAndNext.compareAndSet(nextNode, node, false, false)) {
+                    return true;
+                }
+            }
+        }
+    }
+
+    public boolean remove(T value) {
+        while (true) {
+            Pair pair = find(value);
+            Node currentNode = pair.currentNode;
+            Node nextNode = pair.nextNode;
+            if (nextNode == null || nextNode.value.compareTo(value) != 0) {
+                return false;
+            } else {
+                Node next = nextNode.currentFlagAndNext.getReference();
+                if (!nextNode.currentFlagAndNext.compareAndSet(next, next, false, true)) {
+                    continue;
+                }
+                // No need really delete right now
+                currentNode.currentFlagAndNext.compareAndSet(nextNode, next, false, false);
+                return true;
+            }
+        }
+    }
+
+
+    public boolean contains(T value) {
+        Node node = head.currentFlagAndNext.getReference();
+        while (node.currentFlagAndNext.getReference() != null && node.value.compareTo(value) < 0) {
```

Comment:
= Iteration 2 / 10 =
Actors per thread:
[get(1), put(1), remove(4)]
[remove(5), put(2), get(4), put(3), put(3), put(1)]
[put(5), get(1), remove(5)]
Exception in thread "main" java.lang.IllegalStateException: java.lang.reflect.InvocationTargetException
        at com.devexperts.dxlab.lincheck.LinChecker.lambda$check$0(LinChecker.java:84)
      ...
Caused by: java.lang.NullPointerException
        at LockFreeSetImpl$Node.access$100(LockFreeSetImpl.java:4)
        at LockFreeSetImpl.contains(LockFreeSetImpl.java:65)
        at com.hpcourse.ees.linCheckTester.LockFreeSetCheck.get(LockFreeSetCheck.java:32)
        ... 36 more

Reply:
Fixed.